src/robot/libraries/BuiltIn.py

```diff
@@ -3088,7 +3088,17 @@ def _get_possibly_appended_value(self, initial, new, append):
         if not is_unicode(new):
             new = unic(new)
         if is_truthy(append) and initial:
+
+            if new.startswith('*HTML*'):
+                new = new[6:].lstrip()
+                if initial.startswith('*HTML*'):
+                    return '%s %s' % (initial, new)
+                else:
+                    initial = '*HTML* ' + initial
+            else:
+                new = html_escape(new)
             return '%s %s' % (initial, new)
```

Comment:
Various bigger and smaller issues in this code:

- This helper method is used also by other methods than `set_test_message` and in those usages nothing should be done if message happens to start with `*HTML*`. Should either implement separate helper that handles `*HTML*` or make the current one configurable with something like `handle_html=False`.

- `html_escape(new)` should be one *only* when `initial` is HTML (and when appending). Existing tests for non-HTML appending should be enhanced to contain something like `<message>` to see that they are not altered.

- It's bit hard to follow the logic when some branches return the value and others just modify values that are then returned later. Should handle this consistently. Also, it probably would be best to have

  ```python
  if not (is_truthy(append) and initial):
      return new
  ```
  before handling appending in general. That would avoid one level of nested if/else structures.

- We try to avoid empty rows inside methods. If methods grow so big/complicated that code should be grouped, we rather create helper methods. Helper may not be necessary here, but please remove the empty rows.